libpod/container_api.go

```diff
@@ -745,3 +747,77 @@ func (c *Container) Sync() error {
 
 	return nil
 }
+
+func (c *Container) setupOCIHooks(g *generate.Generator) error {
+	addedHooks := map[string]struct{}{}
+	addHook := func(hook hooks.HookParams) error {
+		// Only add a hook once
+		if _, ok := addedHooks[hook.Hook]; !ok {
+			if err := c.addOCIHook(g, hook); err != nil {
+				return err
+			}
+			addedHooks[hook.Hook] = struct{}{}
+		}
+		return nil
+	}
+	for _, hook := range hooks.Hooks() {
+		logrus.Debugf("SetupOCIHooks", hook)
+		if hook.HasBindMounts && len(c.config.Mounts) > 0 {
+			if err := addHook(hook); err != nil {
+				return err
+			}
+			continue
+		}
+		for _, cmd := range hook.Cmds {
+			match, err := regexp.MatchString(cmd, c.config.Spec.Process.Args[0])
+			if err != nil {
+				logrus.Errorf("Invalid regex %q:%q", cmd, err)
+				continue
+			}
+			if match {
+				if err := addHook(hook); err != nil {
+					return err
+				}
+			}
+		}
+		for _, annotationRegex := range hook.Annotations {
+			for _, annotation := range c.config.Spec.Annotations {
+				match, err := regexp.MatchString(annotationRegex, annotation)
+				if err != nil {
+					logrus.Errorf("Invalid regex %q:%q", annotationRegex, err)
+					continue
+				}
+				if match {
+					if err := addHook(hook); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// addOCIHook look for hooks programs installed in hooksDirPath and add them to spec
+func (c *Container) addOCIHook(g *generate.Generator, hook hooks.HookParams) error {
+	logrus.Debugf("AddOCIHook", hook)
+	for _, stage := range hook.Stage {
+		h := spec.Hook{
+			Path: hook.Hook,
+			Args: append([]string{hook.Hook}, hook.Arguments...),
+			Env:  []string{fmt.Sprintf("stage=%s", stage)},
+		}
+
+		switch stage {
+		case "prestart":
+			g.AddPreStartHook(h)
+
+		case "poststart":
+			g.AddPostStartHook(h)
+
+		case "poststop":
+			g.AddPostStopHook(h)
+		}
+	}
```

Comment:
Should we have  a default with at least a debug statement?

Reply:
Agree, we should handle default in some way. Probably worth erroring, because the hook has not been added to any stage?

Reply:
Good point fixed.